stan/math/rev/fun/matrix_power.hpp

```diff
@@ -130,38 +25,50 @@ class matrix_product_vari : public vari {
  * @throw std::invalid_argument if the exponent is negative or the matrix is not
  * square.
  */
-template <typename EigMat, require_eigen_vt<is_var, EigMat>* = nullptr>
-inline Eigen::Matrix<value_type_t<EigMat>, EigMat::RowsAtCompileTime,
-                     EigMat::ColsAtCompileTime>
-matrix_power(const EigMat& M, const int n) {
-  using T = value_type_t<EigMat>;
-  constexpr int R = EigMat::RowsAtCompileTime;
-  constexpr int C = EigMat::ColsAtCompileTime;
-
-  check_square("matrix_power", "M", M);
-  if (n < 0)
-    invalid_argument("matrix_power", "n", n, "is ", ", but must be >= 0!");
-  if (M.rows() == 0)
-    invalid_argument("matrix_power", "M.rows()", M.rows(), "is ",
-                     ", but must be > 0!");
+template <typename T, require_rev_matrix_t<T>* = nullptr>
+inline plain_type_t<T> matrix_power(const T& M, const int n) {
   const auto& M_ref = to_ref(M);
+  check_square("matrix_power", "M", M);
+  check_nonnegative("matrix_power", "n", n);
   check_finite("matrix_power", "M", M_ref);
-  if (n == 0) {
-    auto* baseVari = new internal::matrix_product_vari_n0<R, C>(M_ref);
-    Eigen::Matrix<var, R, C> Mn(M.rows(), M.cols());
-    Mn.vi() = Eigen::Map<matrix_vi>(baseVari->adjMnRef_, M.rows(), M.cols());
-    return Mn;
-  } else if (n == 1) {
-    auto* baseVari = new internal::matrix_product_vari_n1<R, C>(M_ref);
-    Eigen::Matrix<var, R, C> Mn(M.rows(), M.cols());
-    Mn.vi() = Eigen::Map<matrix_vi>(baseVari->adjMnRef_, M.rows(), M.cols());
-    return Mn;
-  } else {
-    auto* baseVari = new internal::matrix_product_vari<R, C>(M_ref, n);
-    Eigen::Matrix<var, R, C> Mn(M.rows(), M.cols());
-    Mn.vi() = Eigen::Map<matrix_vi>(baseVari->adjMnRef_, M.rows(), M.cols());
-    return Mn;
+
+  if (M.size() == 0)
+    return M;
+
+  size_t N = M.rows();
+
+  if (n == 0)
+    return Eigen::MatrixXd::Identity(N, N);
+
+  if (n == 1)
+    return M_ref;
+
+  arena_t<std::vector<Eigen::MatrixXd>> powers(n + 1);
+  arena_t<plain_type_t<T>> arena_M = M_ref;
+
+  powers[0] = Eigen::MatrixXd::Identity(N, N);
+  powers[1] = value_of(M_ref).eval();
```

Comment:
I think you need to give them a default size when you make the vector like 

```
arena_t<std::vector<Eigen::MatrixXd>> powers(n + 1, Eigen::MatrixXd(N, N));
```

Otherwise the vector holds a bunch of default initialized matrices